### ISSUE TITLE
lockfile updates failing to push when using the amend option

### ISSUE THREAD

**Issue by travi** [greenkeeperio/greenkeeper-lockfile#174]
i've been noticing that my lockfiles are not up to date on several projects lately, but didnt pay close attention since random churn isn't uncommon with the lockfiles quite yet. however, now that i've finally looked into it, i do see failures like [this one](https://travis-ci.com/travi/travis-scaffolder-javascript/builds/75805034#L1907-L1917) in several projects. 

it appears to be related to the force push failing for some reason on projects where i've enabled the option to amend rather than adding an additional commit by setting `GK_LOCK_COMMIT_AMEND`. i thought i've seen that working after [my PR](https://github.com/greenkeeperio/greenkeeper-lockfile/pull/154) to force push was accepted, but i'm having trouble finding any supporting evidence. i must have just accepted the removal of the second commit to mean it was working.

since the output doesn't give much to go on, would you maybe have any suggestions to debug this further? i am really surprised that the `--force-with-lease` was not the appropriate fix in this case.

**Comment by travi** [comment]
it looks like [the push](https://github.com/greenkeeperio/greenkeeper-lockfile/blob/701d5cd05616b7996031f4f92a04a5a2da0904b2/upload.js#L57) limits to just the `HEAD` commit. i'm not very experienced with this approach, but i'm betting this doesn't work with a force-push. 

should the `HEAD:` simply be removed, leaving just the remote branch name? will that cause other issues?

**Comment by travi** [comment]
@finnp sorry to ping you directly, but since you've given feedback on this feature along the way, i'm hoping you have a few thoughts you could provide. 

i'd really like to help move this forward toward a fix, but i'm not quite sure what to try since the error above doesn't give much info. even if i fork and depend on a version with my own guesses, i'd have to wait for a greenkeeper PR to be sent to the project that i updated to even get feedback of whether i'm on the right track. 

i've been generating several new projects with this option set, so i'm wondering if waiting to help fix this is the next move i need to make or if i need to start disabling this option on those projects for now

**Comment by richardson-trevor** [comment]
Since we are adding a new remote [here](https://github.com/travi/greenkeeper-lockfile/blob/c3b3f89559b74b17a49058f07291326e06f52713/upload.js#L55), I wonder if we'd have to fetch the new remote locally before using `force-with-lease` since `force-with-lease` tells it to verify that the local branch sha matches the remote branch sha.

**Comment by richardson-trevor** [comment]
Assuming that is actually the problem, I wonder if something like this would give us the protection of `--force-with-lease` but without first fetching the new remote branches:
```
git push${process.env.GK_LOCK_COMMIT_AMEND ? ` --force-with-lease=${info.branchName}:origin/${info.branchName}` : ''} gk-origin HEAD:${info.branchName}
```

**Comment by guillaumearm** [comment]
I have the same problem on this [CircleCI build](https://circleci.com/gh/guillaumearm/handle-io/4364?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link).

IMO `git push --force-with-lease` fails because there are 2 differents authors (in my case)
Please take a look to [this PR commits](https://github.com/guillaumearm/handle-io/pull/158/commits).
It's an old successful PR (without the AMEND option).
You can see 2 differents greenkeeper bots : `greenkeeper[bot]` and `greenkeeperio-bot`

`--force-with-lease` is used to avoid erase the work of others, maybe it's the cause.

**Comment by travi** [comment]
i tried forking and [making an update](https://github.com/greenkeeperio/greenkeeper-lockfile/compare/master...travi:force-with-lease) to leverage the [suggestion](https://github.com/greenkeeperio/greenkeeper-lockfile/issues/174#issuecomment-400363829) from @richardson-trevor above. it does enable successful pushing of the amended commit, so thanks a ton for that info @richardson-trevor!

however, this results in looping of the builds because `greenkeeper-lockfile` re-runs on the amended commit, pushing another amend, and triggering another build. i havent yet looked into what is missing as a stop condition for this scenario vs the traditional flow with the extra commit, but clearly there is a bit more still outstanding for the amend feature to be complete. 

i'll go ahead and send a PR the change that enables the push, but we really need to get the looping behavior solved as well before the amend feature is actually usable.

**Comment by travi** [comment]
it looks like the `hasLockfileCommit` function under `src/git-helpers.js` simply [checks for a commit with the specific message that is traditionally added by `greenkeeper-lockfile`](https://github.com/greenkeeperio/greenkeeper-lockfile/blob/53305bd2a1b91c5b986bfcf25ccf4b1358a8a6d5/lib/git-helpers.js#L44).

to get the check to work with amends, it could possibly confirm that it has the commit made by the greenkeeper bot, but has both the `package.json` and `package-lock.json` instead of just `package.json`. anyone have other ideas that would be more explicit?

**Comment by janl** [comment]
@travi parsing by commit message was the simplest thing I could get to work, but if you manage to parse any other identifier, like the committer, we could match for that easily too

**Comment by travi** [comment]
I'll try to take a look when i get a chance, but I'm worried that having the amend option enabled without that part fixed will now result in a build lol that will never stop. I probably have a few projects or there where i haven't disabled yet, so this would cause some trouble in it's current state.  

just wondering ifa fix for this might need some additional attention before i can get back to it.

**Comment by travi** [comment]
just to clarify the issue that remains after that PR was merged, i want to keep this piece visible:

> however, this results in looping of the builds because greenkeeper-lockfile re-runs on the amended commit, pushing another amend, and triggering another build.

**Comment by travi** [comment]
i havent verified that this solves the problem yet, but its looking like if we change https://github.com/greenkeeperio/greenkeeper-lockfile/blob/e3dc6881cbaf7c89a082be9cdcb1645081d00f49/lib/git-helpers.js#L44 to something close to 

```js
exec(`git show --oneline --name-only origin/${info.branchName}...${defaultBranch} | grep package-lock.json`)
```

it should get us pretty close to what we're after

**Comment by janl** [comment]
Correction, we also have to grep for `yarn.lock` and `npm-shrinkwrap.json`, I’ll have a PR up shortly.

**Comment by travi** [comment]
thanks so much for taking this over the finish line, @janl 

i'll get it pulled into a few projects to confirm there are no other surprises, but i think this should be the last of them

**Comment by janl** [comment]
Thanks @travi ^5

**Comment by travi** [comment]
@janl since these fixes got merged into what is being released as `next`, would it make sense to cherry pick these changes back onto latest? if it wouldnt introduce too much trouble on your end, it could be helpful to us to separate this from the monorepo changes that are in flight and appear to have a few kinks to work out.

**Comment by travi** [comment]
Fair enough. Thanks for keeping it moving along. I'll give it another shot today with the latest changes

**Comment by travi** [comment]
sorry for not following up before now, but i've been distracted with other things after running into what i thought were still unrelated (to the amend related changes) issues with `next`. however, i upgraded a few other projects to `next` and just confirmed that the amend is working correctly.

notice that there is only 1 commit, but two changed files (including the lockfile), in [this PR](https://github.com/travi/scaffolder-sub-command/pull/87)

**Comment by travi** [comment]
@janl any interest in [reconsidering](https://github.com/greenkeeperio/greenkeeper-lockfile/issues/174#issuecomment-406214214) pulling this change over to the `latest` channel since this is now blocked by https://github.com/greenkeeperio/greenkeeper-lockfile/issues/164#issuecomment-411559411?
